utils/fork-tree/src/lib.rs

```diff
@@ -755,48 +805,64 @@ mod node_implementation {
 				  F: Fn(&H, &H) -> Result<bool, E>,
 				  P: Fn(&V) -> bool,
 		{
-			// stop searching this branch
-			if *number < self.number {
-				return Ok(FindOutcome::Failure(false));
-			}
+			let mut stack: Vec<(&Self, Option<usize>, bool)> = vec![(self, None, true)];
```

Comment:
I didn't really understand everything that's being done here, I think something like this should work:

```rust
		pub fn find_node_index_where<F, P, E>(
			&self,
			hash: &H,
			number: &N,
			is_descendent_of: &F,
			predicate: &P,
		) -> Result<FindOutcome<Vec<usize>>, Error<E>>
		where E: std::error::Error,
			  F: Fn(&H, &H) -> Result<bool, E>,
			  P: Fn(&V) -> bool,
		{
			let mut stack = vec![(self, None, false)];
			let mut indices = Vec::new();

			while let Some((node, index, expanded)) = stack.pop() {
				// if we have already expanded this node we drop the index
				// for the tree path
				if expanded {
					indices.pop();
				}

				// skip to the next node as we have traveled too deep
				if *number < node.number {
					continue;
				}

				// if we haven't expanded this node yet and it has children
				// let's keep traversing in depth
				if !expanded && !node.children.is_empty() {
					// we re-add this node to the stack
					// noting that it has already been expanded
					stack.push((node, index, true));

					// we also need add the index of this node
					// in the tree path
					if let Some(idx) = index {
						indices.push(idx);
					}

					// add all children to the stack
					stack.extend(node.children.iter().enumerate().map(|(i, n)| (n, Some(i), false)));

					continue;
				}

				// we're in the right branch
				if is_descendent_of(&node.hash, hash)? {
					// and this node passes the predicate
					if predicate(&node.data) {
						// we must re-add its index (if any) since it was popped
						// in the beginning of this iteration, and we also need
						// to reverse the tree path
						if let Some(idx) = index {
							indices.push(idx);
							indices.reverse();

							return Ok(FindOutcome::Found(indices));
						} else {
							// if there's no index it means that the found node
							// is `self` so we return an empty tree path
							return Ok(FindOutcome::Found(Vec::new()));
						}
					}
				}
			}

			Ok(FindOutcome::Failure(false))
		}
```